pilot/pkg/config/kube/ingress/controller.go

```diff
@@ -82,6 +82,8 @@ type controller struct {
 	mutex sync.RWMutex
 	// processed ingresses
 	ingresses map[types.NamespacedName]*knetworking.Ingress
+	// serviceStore stores the services that are of port name type, specified by ingresses.
+	serviceStore *serviceStore
```

Comment:
Can you explain why we need a brand new data structure and event handler rather than using the Service kclient already on the ingress `controller`?

Reply:
1.  The `serviceStore` is an indexer used to help us quickly determine whether a service is referenced by the ingress using the port name. 
2. Here why I register event handler for service kclient is that we should detect the service events to determine whether the target service is referenced by the ingress using the port name and notify the `discover server` to re-init virtualservice in pushContext.

I have already used the the Service kclient :).

Reply:
Sorry, was reading on my phone. Sounds good; will take a look